rust_src/src/composite.rs

```diff
@@ -0,0 +1,89 @@
+//! Composite sequence support.
+
+use std::convert::TryInto;
+
+use remacs_macros::lisp_fn;
+
+use crate::{
+    fns::validate_subarray_rust,
+    lisp::LispObject,
+    multibyte::LispStringRef,
+    remacs_sys::{EmacsInt, Fcons, Fput_text_property, Qcomposition},
+};
+
+/// Internal use only.
+///
+/// Compose text between indices START and END of STRING, where
+/// START and END are treated as in `substring'.  Optional 4th
+/// and 5th arguments are COMPONENTS and MODIFICATION-FUNC
+/// for the composition.  See `compose-string' for more details.
+#[lisp_fn(min = "3")]
+pub fn compose_string_internal(
+    string: LispObject,
```

Comment:
Because you are now using `LispObject::from` you can put those types right here into the function signature. The `lisp_fn` machinery will run `foo.into()` on each parameter.

```
pub fn compose_string_internal(
    string: LispStringRef,
    start: EmacsInt,
    end: EmacsInt,
    components: LispObject,
    modification_func: LispObject
) -> LispObject
```